src/gossip/fuzz_table.zig

```diff
@@ -0,0 +1,294 @@
+const EndPoint = network.EndPoint;
+const Packet = sig.net.packet.Packet;
+const PACKET_DATA_SIZE = sig.net.packet.PACKET_DATA_SIZE;
+const GossipMessage = sig.gossip.message.GossipMessage;
+const Ping = sig.gossip.ping_pong.Ping;
+const Pong = sig.gossip.ping_pong.Pong;
+const GossipTable = sig.gossip.GossipTable;
+
+const SignedGossipData = sig.gossip.data.SignedGossipData;
+const GossipData = sig.gossip.data.GossipData;
+const GossipKey = sig.gossip.data.GossipKey;
+const Signature = sig.core.Signature;
+
+const GossipPullFilterSet = sig.gossip.pull_request.GossipPullFilterSet;
+const GossipPullFilter = sig.gossip.pull_request.GossipPullFilter;
+const Hash = sig.core.hash.Hash;
+const ThreadPool = sig.sync.thread_pool.ThreadPool;
+
+const TRIM_INTERVAL = 2 * std.time.ns_per_s;
+const MAX_N_THREADS = 2;
+
+pub fn run(seed: u64, args: *std.process.ArgIterator) !void {
+    const maybe_max_actions_string = args.next();
+    const maybe_max_actions = blk: {
+        if (maybe_max_actions_string) |max_actions_str| {
+            break :blk try std.fmt.parseInt(usize, max_actions_str, 10);
+        } else {
+            break :blk null;
+        }
+    };
+
+    // setup
+    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
+    const allocator = gpa.allocator();
+
+    const logger = Logger.init(allocator, .debug);
+    defer logger.deinit();
+    logger.spawn();
+
+    var prng = std.rand.DefaultPrng.init(seed);
+    const rand = prng.random();
+
+    // init gossip table
+    // MAX: 2 threads
+    const n_threads = @min(@as(u32, @truncate(std.Thread.getCpuCount() catch 1)), MAX_N_THREADS);
+    const thread_pool = try allocator.create(ThreadPool);
+    defer {
+        thread_pool.shutdown();
+        thread_pool.deinit();
+        allocator.destroy(thread_pool);
+    }
+    thread_pool.* = ThreadPool.init(.{
+        .max_threads = n_threads,
+        .stack_size = 2 * 1024 * 1024,
+    });
+
+    const gossip_table = try allocator.create(GossipTable);
+    gossip_table.* = try GossipTable.init(allocator, thread_pool);
+    defer {
+        gossip_table.deinit();
+        allocator.destroy(gossip_table);
+    }
+
+    const Actions = enum { put, get };
+    var put_count: u64 = 0;
+    var get_count: u64 = 0;
+    var total_action_count: u64 = 0;
+    var now: u64 = 0;
+
+    var insertion_times = try std.ArrayList(u64).initCapacity(allocator, 100);
+    defer insertion_times.deinit();
+
+    var pubkeys = try std.ArrayList(Pubkey).initCapacity(allocator, 100);
+    defer pubkeys.deinit();
+
+    var keypairs = try std.ArrayList(KeyPair).initCapacity(allocator, 100);
+    defer keypairs.deinit();
+
+    var signatures = std.AutoArrayHashMap(Pubkey, Signature).init(allocator);
+    defer signatures.deinit();
+
+    var keys = try std.ArrayList(GossipKey).initCapacity(allocator, 100);
+    defer keys.deinit();
+
+    var timer = std.time.Timer.start() catch unreachable;
+
+    var seed_buf: [32]u8 = undefined;
```

Comment:
Any reason to declare this so far away from the site where it's actually used?

Reply:
we only need to define it once - im pretty sure if it was inside the loop the space would be re-allocated on the stack every iteration

Reply:
A 32 byte stack allocation is essentially free, and a good optimizer is going to elide it anyway, I don't think the theoretical benefit outweighs the (albeit small) real mental overhead of declaring this so far away from where it's actually used.

Reply:
Just to back this statement up, here's [a comparison in godbolt](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:zig,selection:(endColumn:38,endLineNumber:13,positionColumn:38,positionLineNumber:13,selectionStartColumn:38,selectionStartLineNumber:13,startColumn:38,startLineNumber:13),source:'const+std+%3D+@import(%22std%22)%3B%0A%0Aexport+fn+foo()+void+%7B%0A++++var+data:+%5B32%5Du8+%3D+undefined%3B%0A++++while+(true)+%7B%0A++++++++@memset(%26data,+1)%3B%0A++++++++std.mem.doNotOptimizeAway(%26data)%3B%0A++++%7D%0A%7D%0A%0Aexport+fn+bar()+void+%7B%0A++++while+(true)+%7B%0A++++++++var+data:+%5B32%5Du8+%3D+undefined%3B%0A++++++++@memset(%26data,+1)%3B%0A++++++++std.mem.doNotOptimizeAway(%26data)%3B%0A++++%7D%0A%7D'),l:'5',n:'1',o:'Zig+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:z0120,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:zig,libs:!(),options:'-OReleaseFast',overrides:!(),selection:(endColumn:1,endLineNumber:18,positionColumn:1,positionLineNumber:18,selectionStartColumn:1,selectionStartLineNumber:17,startColumn:1,startLineNumber:17),source:1),l:'5',n:'0',o:'+zig+0.12.0+(Editor+%231)',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4).